### ISSUE TITLE
Highpass filtering preprocessing step during training

### ISSUE THREAD

**Issue by alecgunny** [ML4GW/aframe#158]
#153 originally included a highpass filter implementation using `torchaudio` that looked like this:
```python
import math

import torch
from torchaudio.functional import lfilter

from bbhnet.data.transforms.transform import Transform


class HighpassFilter(Transform):
    def __init__(
        self, cutoff: float, sample_rate: float, Q: float = 1 / math.sqrt(2)
    ) -> None:
        super().__init__()
        w0 = 2 * math.pi * cutoff / sample_rate
        alpha = math.sin(w0) / 2.0 / Q

        b0 = (1 + math.cos(w0)) / 2
        b1 = -1 - math.cos(w0)
        b2 = b0
        a0 = 1 + alpha
        a1 = -2 * math.cos(w0)
        a2 = 1 - alpha

        self.a_coeffs = self.add_parameter([a0, a1, a2])
        self.b_coeffs = self.add_parameter([b0, b1, b2])

    def forward(self, X: torch.Tensor) -> torch.Tensor:
        return lfilter(X, self.a_coeffs, self.b_coeffs)
```

but there were a number of issues with getting it implemented properly that caused it to be taken out. Documenting them here for future investigation on how to fix them, and if including such a filter improves model robustness:

- ONNX doesn't support the torch operation used to implement the `lfilter`. This would appear to be nontrivial, as in [this thread](https://github.com/pytorch/audio/issues/982) about supporting the `stft` op such a workaround is never suggested. If we couldn't support it on the server, such an op would have to be implemented on the client side at inference time, which would introduce a torch dependency that would be really undesirable. The alternative is to filter using scipy/gwpy/etc., but since we need to filter _after_ injection gets done during training (we could in principle filter the glitches once up front), this would mean transferring the GPU tensor of injected background back to CPU, transferring to numpy, filtering, transferring back to Torch, then back to the GPU, which would likely slow down training significantly. The last alternative is to implement a filtering operation using regular torch operations in a loop (`torchaudio` already has [such an implementation](https://github.com/pytorch/audio/blob/539476e246270b0be5614156cc2a5b0d2de98bcf/torchaudio/functional/filtering.py#L923)), but this would likely also be extremely slow.
- Even without the workarounds described above which would slow this op down, it's already the bottleneck during training. The issue is that any filter with an IIR component will be inherently recursive, limiting the amount of parallelism you can take advantage of on an accelerator. For reference, this reduces our training throughput by about a factor of 2 as it stands.
- This implementation is probably not ideal at the moment, I'm sure you'd want to do some windowing/edge slicing (akin to how we do things during whitening) to keep unwanted frequency components from leaking in. I'm sure there are some tricks in `gwpy`'s filtering implementations that could be deployed here to improve things.
- Building this environment is actually non-trivial. Basically, if you try to include `torchaudio` as a Poetry dependency in the `bbhnet.data` project, the CPU-only builds used during testing will fail because `torchaudio` will look for GPU kernels that aren't there (because it's being installed in conda environment that has `cudatoolkit` installed from the repo-wide `environment.yaml`). This can be fixed by adding `torchaudio` to the main `environment.yaml`, but then the `bbhnet.architectures` library would need to add a `poetry.toml` to move from a poetry-managed environment to a conda-managed one (since it would depend on `bbhnet.data.transforms` to implement the `Preprocessor`). The alternative here is to host the `Preprocessor` in `bbhnet.data`, which could potentially make sense, but again is just something that would require more thought than #153 was prepared to handle.

**Comment by alecgunny** [comment]
First point is outlined in #154

**Comment by alecgunny** [comment]
Marking as a good first issue mostly for someone to look at the effect of highpass filtering on the network's ability to learn, even if the implementation is suboptimal. If this work shows that this is worth investing time in doing correctly, then we can start digging into the more complicated work of making this efficient.

**Comment by EthanMarx** [comment]
Yeah this crossed my mind at some point before. I'm not sure if or how this differs from calling `.highpass` and then `.whiten` without a highpass. In any case, you're definitely right that there is high passing being done since our training kernels look good.
